feat: 듀오 삭제하기

inseonbak opened the pull request:
## ✨ 담당 파트
- 듀오 삭제

## 🔎 작업 상세 내용
- 듀오 생성자 혹은 매니저가 삭제 가능한 로직 생성

## ✅ 테스트 코드 작성 및 기능 테스트 여부
- [] 테스트 코드 작성
- [] 기능 테스트 여부

## ➕ 이슈 링크
- #1

Commit:
feat(DuoController): Controller #1

- 삭제 로직 생성

Commit:
feat(DuoService): Service #1

-듀오 생성자 혹은 매니저일 때 삭제 가능한 로직 생성

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/controller/DuoController.java:
```diff
@@ -1,4 +1,28 @@
+                          ) {
+        // Duo 삭제 로직
+        duoService.deleteDuoById(duoId,authMember);
+    }
```
1. 저희는 **사용자의 HttpRequest에 대한 응답 데이터**를 **HttpHeader와 HttpBody, Status Code에 담아서 반환**할 거니까, **HTTP 응답을 제어**할 수 있는 **ResponseEntity**를 사용하는 건 어떨까요?

- 반환하고자 한 값이 void였으면, ResponseEntity를 사용하면 반환 값을 void 대신 `ResponseEntity<Void>` 로 표현할 수 있겠네요~
**ResponseEntity**를 사용하면 **Body 타입**은 **Wrapper 클래스**로 지정되어야 하니까요~
<br>

2. 삭제가 정상적으로 진행되면, DB에서 삭제 **요청이 성공적으로 실행**되었지만 **응답 데이터를 반환할 필요가 없다**는 의미를 가지는 **HTTP Status Code 204**번을 사용하는 건 어떨까요?

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/service/DuoService.java:
```diff
@@ -1,7 +1,35 @@
+        }
+
+        duoRepository.deleteById(duoId);
+    }
```
- AuthMember 를 사용하지 않으면, Controller에서 authMember 의 memberId 를 추출해서 Service에 넘겨주면 어떨까요?

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/service/DuoService.java:
```diff
@@ -1,7 +1,35 @@
+                () -> new IllegalArgumentException("듀오가 존재하지 않습니다.")
+        );
+
+        if (!authMember.getMemberId().equals(existDuo.getMemberId())) {
+            if (!authMember.getRole().equals(UserRole.ADMIN)) {
```
- 해당 코드는 가독성이 떨어지는 것 같습니다ㅠㅠ

```suggestion
        if ( authMember.getRole().equals(UserRole.MEMBER) &&
            (authMember.getMemberId().equals(existDuo.getMemberId) )
```
-> 사용자이지만, 해당 듀오 신청자가 아니다

이렇게 변경해보는 건 어떨까요? 좀 더 직관적이지 않나요? :)

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/service/DuoService.java:
```diff
@@ -1,7 +1,35 @@
+            }
+        }
+
+        duoRepository.deleteById(duoId);
```
- Hard Delete의 삭제 방식도 좋은 방식이지만, 저희가 듀오 삭제할 때 **Soft Delete** 방식으로 설계하였지요~

- Soft Delete 하기 위해서 deleted_at 컬럼을 만들어 주었으니, **deleted_at에 삭제하는 현재 시간**을 넣어주면 될 것 같습니다~

for2ver0 reviewed (changes requested):
리뷰 확인 부탁드립니다~

Commit:
test(DuoServiceTest): ServiceTest #1

-삭제가 정상적으로 이루어진 경우 (내가 만든 듀오를 삭제할 경우)
-삭제가 정상적으로 이루어진 경우 (매니저가 삭제할 경우)
-삭제가 정상적으로 이루어지지 않음(권한없음)

yokxim2 reviewed (approved):
고생하셨습니다!

for2ver0 commented on src/test/java/com/summoner/lolhaeduo/domain/duo/service/DuoServiceTest.java:
```diff
@@ -0,0 +1,91 @@
+
+        AuthMember authMember = new AuthMember(memberId, "user1", UserRole.MEMBER);
+
+        // when : 이걸 실행했을 때
+        duoService.deleteDuoById(duo.getId(), authMember);
+
+        // then : 이 값이 나와야한다.
+        Optional<Duo> deletedDuo = duoRepository.findById(duo.getId());
+        Assertions.assertTrue(deletedDuo.isEmpty(), "Duo가 삭제되어야 합니다.");
```
- 현재 로직에서는 올바른 테스트 코드인 것 같습니다 👍 
- 하지만 Soft Delete를 적용하면, 이 부분은 수정되어야 할 듯 합니다.

for2ver0 commented on src/test/java/com/summoner/lolhaeduo/domain/duo/service/DuoServiceTest.java:
```diff
@@ -0,0 +1,91 @@
+        duoRepository.save(duo);
+
+        AuthMember authMember = new AuthMember(2L,"user1", UserRole.ADMIN);
+        //when
+        duoService.deleteDuoById(duo.getId(), authMember);
+        //then
+        Optional<Duo> deletedDuo = duoRepository.findById(duo.getId());
+        Assertions.assertTrue(deletedDuo.isEmpty(), "Duo가 삭제되어야 합니다");
```
- 현재 로직에서는 올바른 테스트 코드인 것 같습니다 👍
- 하지만 Soft Delete를 적용하면, 이 부분 수정되어야 할 듯 합니다.

for2ver0 commented on src/test/java/com/summoner/lolhaeduo/domain/duo/service/DuoServiceTest.java:
```diff
@@ -0,0 +1,91 @@
+
+        AuthMember unauthorizedMember = new AuthMember(unauthorizedMemberId, "user2", UserRole.MEMBER);
+
+        //when
+        Assertions.assertThrows(
+                IllegalArgumentException.class, () -> duoService.deleteDuoById(duo.getId(), unauthorizedMember));
+
+        //then
+        Optional<Duo> deletedDuo = duoRepository.findById(duo.getId());
+        Assertions.assertTrue(deletedDuo.isPresent(), "Duo가 삭제되지 않았습니다."); // 데이터가 유지됨을 확인
```
- 현재 로직에서는 너무 좋은 테스트 코드인 것 같습니다 👍 
- 하지만 deleted_at 컬럼을 이용한 Soft Delete를 적용한다면, 이 부분도 수정되어야 할 것 같습니다~

for2ver0 reviewed (changes requested):
테스트 코드까지 작성하시느라 너무 수고하셨습니다!!
리뷰 확인 부탁드립니다~

Commit:
entity(Duo): DuoEntity #1

-delete 메서드 추가

Commit:
Controller(DuoController): DuoController #1

-ResponseEntity<void> 사용
-Http 상태 코드 204 반환

Commit:
Repository(DuoRepository): DuoRepository #1

-@Repository 추가

Commit:
Service(DuoService): DuoService #1

-삭제 시간 기록
-엔티티를 업데이트하여 삭제 시간 저장

Commit:
ServiceTest(DuoServiceTest): DuoServiceTest #1

-soft delete test
